blog/_posts/2015-12-2-includeos-is-now-open-source.md

```diff
@@ -0,0 +1,39 @@
+
+## It's a bit like a JVM, but for x86 C++
+A Java Virtual Machine is a portable language runtime environment. Java is portable across hardware architectures and operating systems because it uses a common instruction set. Once you've started a Java program, you can't log into it (unless your program itself provides the facilities), and you can't boot up any other programs inside it.
+
+IncludeOS is like a safe language runtime for C++ programs, compiled into the x86 instruction set. This has the obvious advantage of removing one layer of abstraction, compared to Java: with hardware virtualization the code will execute directly on the CPU. Like with a Java program and other unikernels, an IncludeOS service is a program, with the language runtime attached. It's called IncludeOS because you start by saying `#include <os>` and whatever your service needs from the operating system will be included into the virtual machine. 
+
+Think about the original reasons for creating Java; one main reason was definitely portability. With pervasive hardware virtualization you can now get similar portability for compiled C++ programs. We're currently only portable across x86 operating systems, but with virtualbox, you can run the same binary on windows, Linux and Mac. What's more, "Hello World" in IncludeOS on KVM has about *a third of the memory footprint* of "Hello world" in Java, running directly on the server, even when we take the whole Qemu process into account (8.45 MB v.s 28.29 MB)
+
+From this interesting premise, rather than working towards running a JVM or supporting another high-level language, we want to focus on creating an intuitive and modern C++ hardware API, with minimal overhead. But of course, if somebody wants to use IncludeOS as a platform for supporting high level language runtimes, that would be great too. Before that can happen though, we'll have to get a stable API. We're currently on 0.7.0, meaning that anything can change at any time.
+
+## Networking from scratch
+One thing you'll notice is that our network stack is currently being written from scratch. Are we crazy? Possibly. We obviously considered - and wanted to - port the stack from somewhere else, but not being intimately familiar with any existing kernel, I found that to be even harder than just doing it from scratch. I got a lot of help from reading the nicely annotated and cross-linked [SanOS source code](http://www.jbox.dk/sanos/source/), and once the virtio network driver was working, the rest wasn't nearly as hard as one might think - at least up to now. Clearly it will take time to make it complete, but the efforts it took to get from a driver, up to rudimentary TCP was really insignificant compared to the challenges we've faced when trying to integrate third-party libraries such as newlib (Standard C) and libc++ (Standard C++). Getting to read RFC's and implementing the TCP handshake from scratch has really been a blast - a highly recommended exercise.
```

Comment:
This would actually be really interesting reading :)

Perhaps the basis of a future CACM article?